File: packages/runtime-web-three/src/gltfSceneHandles.ts
import type { IGltfNodeHandleIr, IGltfNodeOperationIr, IGltfSceneHandlesIr, IGltfSceneMetadataIr, IGltfSceneNodeIr, ITransformComponent } from "@threenative/ir";

export interface IGltfSceneHandleObservation {
  after?: {
    material?: string;
    transform?: ITransformComponent;
    visible?: boolean;
  };
  before?: {
    material?: string;
    transform?: ITransformComponent;
    visible?: boolean;
  };
  extras?: unknown;
  handle: string;
  nodePath?: string;
  operation: IGltfNodeOperationIr["kind"] | "resolve";
  status: "applied" | "deferred" | "missing";
}

interface INodeState {
  material?: string;
  node: IGltfSceneNodeIr;
  transform: ITransformComponent;
  visible: boolean;
}

export function applyGltfSceneHandleOperations(
  metadata: IGltfSceneMetadataIr,
  handles: IGltfSceneHandlesIr,
  options: { barrierReady?: boolean } = {},
): IGltfSceneHandleObservation[] {
  const barrierReady = options.barrierReady ?? true;
  const states = new Map<string, INodeState>();
  const observations: IGltfSceneHandleObservation[] = [];
  const handlesById = new Map(handles.handles.map((handle) => [handle.id, handle]));

  for (const handle of [...handles.handles].sort((left, right) => left.id.localeCompare(right.id))) {
    const node = resolveNode(metadata, handle);
    if (node === undefined) {
      observations.push({ handle: handle.id, operation: "resolve", status: "missing" });
      continue;
    }
    states.set(handle.id, {
      material: node.materials?.[0],
      node,
      transform: fromNodeTransform(node),
      visible: true,
    });
  }

  for (const operation of [...handles.operations].sort((left, right) => operationSortKey(left).localeCompare(operationSortKey(right)))) {
    const handle = handlesById.get(operation.handle);
    const state = states.get(operation.handle);
    if (handle === undefined || state === undefined) {
      observations.push({ handle: operation.handle, operation: operation.kind, status: "missing" });
      continue;
    }
    if (!barrierReady) {
      observations.push({ handle: operation.handle, nodePath: state.node.path, operation: operation.kind, status: "deferred" });
      continue;
    }
    observations.push(applyOperation(state, operation));
  }

  return observations.sort((left, right) => `${left.handle}:${left.operation}`.localeCompare(`${right.handle}:${right.operation}`));
}

function applyOperation(state: INodeState, operation: IGltfNodeOperationIr): IGltfSceneHandleObservation {
  if (operation.kind === "visibility") {
    const before = state.visible;
    state.visible = operation.visible;
    return {
      after: { visible: state.visible },
      before: { visible: before },
      handle: operation.handle,
      nodePath: state.node.path,
      operation: operation.kind,
      status: "applied",
    };
  }
  if (operation.kind === "transform") {
    const before = state.transform;
    state.transform = { ...state.transform, ...operation.transform };
    return {
      after: { transform: state.transform },
      before: { transform: before },
      handle: operation.handle,
      nodePath: state.node.path,
      operation: operation.kind,
      status: "applied",
    };
  }
  if (operation.kind === "material") {
    const before = state.material;
    state.material = operation.material;
    return {
      after: { material: state.material },
      before: before === undefined ? {} : { material: before },
      handle: operation.handle,
      nodePath: state.node.path,
      operation: operation.kind,
      status: "applied",
    };
  }
  return {
    extras: state.node.extras,
    handle: operation.handle,
    nodePath: state.node.path,
    operation: operation.kind,
    status: "applied",
  };
}

function resolveNode(metadata: IGltfSceneMetadataIr, handle: IGltfNodeHandleIr): IGltfSceneNodeIr | undefined {
  const asset = metadata.assets.find((item) => item.assetId === handle.assetId);
  if (asset === undefined) {
    return undefined;
  }
  if (handle.nodePath !== undefined) {
    return asset.nodes.find((node) => node.path === handle.nodePath);
  }
  return asset.nodes.filter((node) => node.name === handle.nodeName).length === 1
    ? asset.nodes.find((node) => node.name === handle.nodeName)
    : undefined;
}

function fromNodeTransform(node: IGltfSceneNodeIr): ITransformComponent {
  return {
    ...(node.transform?.translation === undefined ? {} : { position: node.transform.translation }),
    ...(node.transform?.rotation === undefined ? {} : { rotation: node.transform.rotation }),
    ...(node.transform?.scale === undefined ? {} : { scale: node.transform.scale }),
  };
}

function operationSortKey(operation: IGltfNodeOperationIr): string {
  return `${operation.handle}:${operation.kind}`;
}

File: runtime-bevy/crates/threenative_runtime/src/gltf_scene_handles.rs
use serde::Serialize;

#[derive(Clone, Debug, PartialEq)]
pub struct NativeGltfSceneNode {
    pub asset_id: String,
    pub extras: Option<serde_json::Value>,
    pub material: Option<String>,
    pub name: Option<String>,
    pub path: String,
    pub transform: NativeGltfTransform,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeGltfTransform {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub position: Option<[f32; 3]>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rotation: Option<[f32; 4]>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scale: Option<[f32; 3]>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct NativeGltfNodeHandle {
    pub asset_id: String,
    pub id: String,
    pub instance_id: String,
    pub node_name: Option<String>,
    pub node_path: Option<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum NativeGltfNodeOperation {
    ExtrasLookup { handle: String },
    Material { handle: String, material: String },
    Transform { handle: String, transform: NativeGltfTransform },
    Visibility { handle: String, visible: bool },
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeGltfSceneHandleObservation {
    pub handle: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub node_path: Option<String>,
    pub operation: String,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub before: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub after: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extras: Option<serde_json::Value>,
}

pub fn apply_gltf_scene_handle_operations(
    nodes: &[NativeGltfSceneNode],
    handles: &[NativeGltfNodeHandle],
    operations: &[NativeGltfNodeOperation],
    barrier_ready: bool,
) -> Vec<NativeGltfSceneHandleObservation> {
    let mut sorted_handles = handles.to_vec();
    sorted_handles.sort_by(|left, right| left.id.cmp(&right.id));
    let mut states: Vec<(String, NativeGltfSceneNode, bool)> = sorted_handles
        .iter()
        .filter_map(|handle| {
            resolve_node(nodes, handle).map(|node| (handle.id.clone(), node.clone(), true))
        })
        .collect();
    let mut sorted_operations = operations.to_vec();
    sorted_operations.sort_by(|left, right| operation_sort_key(left).cmp(&operation_sort_key(right)));
    let mut observations = Vec::new();
    for operation in sorted_operations {
        let handle_id = operation_handle(&operation);
        let Some((_id, node, visible)) = states.iter_mut().find(|(id, _node, _visible)| id == handle_id) else {
            observations.push(NativeGltfSceneHandleObservation {
                after: None,
                before: None,
                extras: None,
                handle: handle_id.to_owned(),
                node_path: None,
                operation: operation_kind(&operation),
                status: "missing".to_owned(),
            });
            continue;
        };
        if !barrier_ready {
            observations.push(NativeGltfSceneHandleObservation {
                after: None,
                before: None,
                extras: None,
                handle: handle_id.to_owned(),
                node_path: Some(node.path.clone()),
                operation: operation_kind(&operation),
                status: "deferred".to_owned(),
            });
            continue;
        }
        observations.push(apply_operation(node, visible, &operation));
    }
    observations.sort_by(|left, right| {
        format!("{}:{}", left.handle, left.operation).cmp(&format!("{}:{}", right.handle, right.operation))
    });
    observations
}

fn apply_operation(
    node: &mut NativeGltfSceneNode,
    visible: &mut bool,
    operation: &NativeGltfNodeOperation,
) -> NativeGltfSceneHandleObservation {
    match operation {
        NativeGltfNodeOperation::Transform { handle, transform } => {
            let before = serde_json::json!({ "transform": node.transform });
            node.transform = merge_transform(&node.transform, transform);
            NativeGltfSceneHandleObservation {
                after: Some(serde_json::json!({ "transform": node.transform })),
                before: Some(before),
                extras: None,
                handle: handle.clone(),
                node_path: Some(node.path.clone()),
                operation: "transform".to_owned(),
                status: "applied".to_owned(),
            }
        }
        NativeGltfNodeOperation::Visibility { handle, visible: next } => {
            let before = *visible;
            *visible = *next;
            NativeGltfSceneHandleObservation {
                after: Some(serde_json::json!({ "visible": visible })),
                before: Some(serde_json::json!({ "visible": before })),
                extras: None,
                handle: handle.clone(),
                node_path: Some(node.path.clone()),
                operation: "visibility".to_owned(),
                status: "applied".to_owned(),
            }
        }
        NativeGltfNodeOperation::Material { handle, material } => {
            let before = node.material.clone();
            node.material = Some(material.clone());
            NativeGltfSceneHandleObservation {
                after: Some(serde_json::json!({ "material": node.material })),
                before: Some(serde_json::json!({ "material": before })),
                extras: None,
                handle: handle.clone(),
                node_path: Some(node.path.clone()),
                operation: "material".to_owned(),
                status: "applied".to_owned(),
            }
        }
        NativeGltfNodeOperation::ExtrasLookup { handle } => NativeGltfSceneHandleObservation {
            after: None,
            before: None,
            extras: node.extras.clone(),
            handle: handle.clone(),
            node_path: Some(node.path.clone()),
            operation: "extrasLookup".to_owned(),
            status: "applied".to_owned(),
        },
    }
}

fn resolve_node<'a>(
    nodes: &'a [NativeGltfSceneNode],
    handle: &NativeGltfNodeHandle,
) -> Option<&'a NativeGltfSceneNode> {
    let matches: Vec<_> = nodes
        .iter()
        .filter(|node| node.asset_id == handle.asset_id)
        .filter(|node| {
            handle
                .node_path
                .as_ref()
                .map(|path| node.path == *path)
                .unwrap_or_else(|| node.name.as_ref() == handle.node_name.as_ref())
        })
        .collect();
    if matches.len() == 1 { Some(matches[0]) } else { None }
}

fn merge_transform(
    before: &NativeGltfTransform,
    update: &NativeGltfTransform,
) -> NativeGltfTransform {
    NativeGltfTransform {
        position: update.position.or(before.position),
        rotation: update.rotation.or(before.rotation),
        scale: update.scale.or(before.scale),
    }
}

fn operation_handle(operation: &NativeGltfNodeOperation) -> &str {
    match operation {
        NativeGltfNodeOperation::ExtrasLookup { handle }
        | NativeGltfNodeOperation::Material { handle, .. }
        | NativeGltfNodeOperation::Transform { handle, .. }
        | NativeGltfNodeOperation::Visibility { handle, .. } => handle,
    }
}

fn operation_kind(operation: &NativeGltfNodeOperation) -> String {
    match operation {
        NativeGltfNodeOperation::ExtrasLookup { .. } => "extrasLookup",
        NativeGltfNodeOperation::Material { .. } => "material",
        NativeGltfNodeOperation::Transform { .. } => "transform",
        NativeGltfNodeOperation::Visibility { .. } => "visibility",
    }
    .to_owned()
}

fn operation_sort_key(operation: &NativeGltfNodeOperation) -> String {
    format!("{}:{}", operation_handle(operation), operation_kind(operation))
}
